worker-workflow/src/main/java/com/github/cafdataprocessing/workflow/WorkflowSettingsRetriever.java

```diff
@@ -0,0 +1,264 @@
+        tenantConfigApi = new TenantConfigurationApi(apiClient);
+        repositoryConfigApi = new RepositoryConfigurationApi(apiClient);
+    }
+
+    public void addCustomWorkflowConfig(final WorkflowSettings requiredConfig, final Document document)
+        throws ApiException, DocumentWorkerTransientException
+    {
+        final String tenantId = document.getCustomData("tenantId");
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(requiredConfig);
+        final Map<String, Map<String, String>> settings = new HashMap<>();
+        settings.put("task", processTaskConfigs(document, requiredConfig.getTaskSettings()));
+        settings.put("repository", processRepositoryConfigs(document, tenantId, requiredConfig.getRepositorySettings()));
+        settings.put("tenant", processTenantConfigs(tenantId, requiredConfig.getTenantSettings()));
+        document.getField("CAF_WORKFLOW_SETTINGS").set(gson.toJson(settings));
+        document.getTask().getResponse().getCustomData().put("CAF_WORKFLOW_SETTINGS", gson.toJson(settings));
+    }
+
+    private Map<String, String> processTenantConfigs(final String tenantId, final List<String> configs)
+        throws ApiException, DocumentWorkerTransientException
+    {
+        if (configs == null) {
+            return null;
+        }
+        final Map<String, String> customConfigs = new HashMap<>();
+        for (final String config : configs) {
+            try {
+                customConfigs.put(config,
+                                  settingsCache.get(
+                                      new WorkflowSettingsTenantCacheKey(tenantId, config)));
+            } catch (final ExecutionException ex) {
+                if (ex.getCause() instanceof ApiException) {
+                    throw (ApiException) ex.getCause();
+                } else if (ex.getCause() instanceof DocumentWorkerTransientException) {
+                    throw (DocumentWorkerTransientException) ex.getCause();
+                } else {
+                    throw new RuntimeException(ex);
+                }
+            }
+        }
+        return customConfigs;
+    }
+
+    private Map<String, String> processRepositoryConfigs(final Document document, final String tenantId,
+                                                         final Map<String, RepoConfigSource> configs) throws ApiException,
+                                                                                                             DocumentWorkerTransientException
+    {
+        final Map<String, String> customConfigs = new HashMap<>();
+        String repositoryKey = null;
+        for (final Map.Entry<String, RepoConfigSource> config : configs.entrySet()) {
+            switch (config.getValue().getSource()) {
+                case FIELD:
+                    repositoryKey = document.getField(
+                        config.getValue().getKey()).getValues().stream().findFirst().get().getStringValue();
```

Comment:
result of findFirst is not checked before get(). Need better error handling like the CUSTOMDATA case below.